typed-th-in-ghc-9/article.md

```diff
@@ -0,0 +1,205 @@
+
+The pattern is pretty similar to the workflow with `Code`, albeit with `Splice` now.
+
+Make sure to import the appropriate definitions first:
+
+```hs
+import Language.Haskell.TH.Syntax (Q)
+import Language.Haskell.TH.Syntax.Compat (Splice, examineSplice, liftSplice)
+```
+
+And now just replace `Code` with `Splice`:
+
+```diff
+-logLevelFromFlag :: Code Q LogLevel
+-logLevelFromFlag = liftCode $ do  -- liftSplice will turn 'Q' into 'Code'
++logLevelFromFlag :: Splice Q LogLevel
++logLevelFromFlag = liftSplice $ do  -- liftSplice will turn 'Q' into 'Splice'
+  flag <- liftIO $ lookupEnv "LOG_LEVEL"
+  case flag of
+-    -- examineCode will turn 'Code' into 'Q'
+-    Nothing -> examineCode [|| Production ||]  -- We default to Production if the flag is unset
++    -- examineSplice will turn 'Splice' into 'Q'
++    Nothing -> examineSplice [|| Production ||]  -- We default to Production if the flag is unset
+    Just level -> case level of
+-      "DEVELOPMENT" -> examineCode [|| Development ||]
+-      "PRODUCTION" -> examineCode [|| Production ||]
++      "DEVELOPMENT" -> examineSplice [|| Development ||]
++      "PRODUCTION" -> examineSplice [|| Production ||]
+      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
+```
+
+And that's it! `Splice m a` is defined in `th-compat` as `m (TExp a)` in GHC 8 and `Code m a` in GHC 9, and this is why this code works. In addition, functions like `liftSplice` and `examineSplice` will either be defined `liftCode` and `examineCode` (in GHC 9) or `id` (in GHC 8).
+
+## Parentheses in splices
+
+Just one more thing before we conclude this article. You may see surprising behavior regarding the usage of parentheses in splices for both untyped and typed Template Haskell depending on whether you are using GHC 8 or GHC 9.
+
+For example, if we had imported `logLevelFromFlag` qualified, then you'd write `$$(TH.logLevelFromFlag)` in GHC 8, otherwise you'd get a parser error:
+
+```hs
+>>> $$TH.logLevelFromFlag
+
+<interactive>:200:1: error: parse error on input ‘$$’
+```
+
+But it works in GHC 9:
+
+```hs
+>>> $$TH.logLevelFromFlag
+Production
+```
+
+## Conclusion
+
+In this post, we've seen the changes introduced in GHC 9 regarding Typed Template Haskell. We've learned about two different ways to migrate typed Template Haskell code from GHC 8 to GHC 9 and analyzed their differences.
+
+Using `th-compat` has a big advantage in being compatible with two compilers, besides having an interface similar to the one in GHC 9. On the other hand, if supporting GHC 8 is not necessary, it becomes an extra dependency in your project that's potentially unfamiliar to many users. The choice of which strategy to use will ultimately depend on your specific needs, but we hope to have shed light on their pros and cons.
```

Comment:
> with two compilers

meh

Reply:
I changed to GHC versions, I hope it's better.